File: packages/coterie-py/src/coterie/modes/tournament.py
"""`tournament` mode — N-way bracket.

    START -> planner -> [participant_1 || ... || participant_N] -> bracket_judge -> END

v0.1: single round, N-way pairwise judging. Multi-round elimination brackets
(power-of-two participants, recursive halving) is v0.1.x. The architecture
already supports it — `bracket_judge` would return a `next_round` field instead
of `winner` and a loop edge would re-enter `bracket_judge` until one remains.
"""

from langgraph.graph import END, START, StateGraph

from coterie.core.compile import compile_with_interrupts
from coterie.core.executor import AdapterExecutor
from coterie.core.llm.base import LLMClient
from coterie.core.registry import register_mode
from coterie.core.state import CoterieState
from coterie.nodes.agent_runner import make_agent_runner
from coterie.nodes.bracket import make_bracket_judge_node
from coterie.nodes.planner import make_planner_node


@register_mode("tournament")
def build(
    *,
    workdir: str,
    executor: AdapterExecutor,
    config: dict,
    judge_llm: LLMClient | None = None,
    **_,
):
    tournament = config.get("tournament") or {}
    participants = tournament.get("participants") or []

    g: StateGraph = StateGraph(CoterieState)
    g.add_node("planner", make_planner_node())
    for pid in participants:
        g.add_node(
            f"tparticipant_{pid}",
            make_agent_runner(
                role="tournament-participant",
                workdir=workdir,
                executor=executor,
                agent_id=pid,
            ),
        )
    g.add_node("bracket_judge", make_bracket_judge_node(judge_llm))

    g.add_edge(START, "planner")
    for pid in participants:
        g.add_edge("planner", f"tparticipant_{pid}")
        g.add_edge(f"tparticipant_{pid}", "bracket_judge")
    g.add_edge("bracket_judge", END)

    return compile_with_interrupts(g, config)

File: packages/coterie-js/src/modes/tournament.ts
/** Mode "tournament": N agents attempt the task in parallel; a bracket judge
 *  (nodes/bracket.ts) ranks and eliminates per round until one winner remains.
 *  Registered into MODE_REGISTRY at import; built per turn by graph.ts. */

import { END, START, StateGraph } from "@langchain/langgraph";

import type { AdapterExecutor } from "../core/executor.js";
import { CoterieStateAnnotation } from "../core/annotation.js";
import { compileGraph } from "../core/compile.js";
import { registerMode } from "../core/registry.js";
import type { CoterieState } from "../core/state.js";
import type { ModeBuildOpts } from "../core/types.js";
import { makeAgentRunner } from "../nodes/agentRunner.js";
import { makeBracketJudgeNode } from "../nodes/bracket.js";
import { makePlannerNode } from "../nodes/planner.js";

function makeEliminatedAwareParticipant(agent_id: string, workdir: string, executor: AdapterExecutor) {
  const inner = makeAgentRunner({ role: "tournament-participant", workdir, executor, agent_id });
  return async (state: CoterieState) => {
    const eliminated = (state.mode_state?.eliminated_participants ?? []) as string[];
    if (eliminated.includes(agent_id)) return {};
    return inner(state);
  };
}

export function build(opts: ModeBuildOpts) {
  const { workdir, executor, config, judge_llm } = opts;
  const tournament = config.tournament ?? {};
  const participants = (tournament.participants ?? (config.agents as any[]).map((a) => a.id)) as string[];
  const multiRound = (tournament.rounds ?? 1) > 1;

  let g: any = new StateGraph(CoterieStateAnnotation)
    .addNode("planner", makePlannerNode())
    .addNode("bracket_judge", makeBracketJudgeNode(judge_llm ?? null));

  for (const pid of participants) {
    g = g.addNode(`tparticipant_${pid}`, makeEliminatedAwareParticipant(pid, workdir, executor));
  }

  g = g.addEdge(START, "planner");
  for (const pid of participants) {
    g = g.addEdge("planner", `tparticipant_${pid}`).addEdge(`tparticipant_${pid}`, "bracket_judge");
  }

  if (multiRound) {
    const targets: Record<string, string> = { [END]: END };
    for (const pid of participants) targets[`tparticipant_${pid}`] = `tparticipant_${pid}`;
    g = g.addConditionalEdges(
      "bracket_judge",
      (state: any) => {
        if (state.status === "done") return END;
        return participants.map((p) => `tparticipant_${p}`);
      },
      targets,
    );
  } else {
    g = g.addEdge("bracket_judge", END);
  }

  return compileGraph(g, config);
}

registerMode("tournament", build);
